src/coreclr/nativeaot/Runtime/unix/UnixNativeCodeManager.cpp

```diff
@@ -653,11 +653,26 @@ bool UnixNativeCodeManager::GetReturnAddressHijackInfo(MethodInfo *    pMethodIn
     GcInfoDecoderFlags flags = DECODE_RETURN_KIND;
 #if defined(TARGET_ARM) || defined(TARGET_ARM64)
     flags = (GcInfoDecoderFlags)(flags | DECODE_HAS_TAILCALLS);
+#elif defined(TARGET_AMD64) && defined(TARGET_OSX)
+    flags = (GcInfoDecoderFlags)(flags | DECODE_PROLOG_LENGTH);
 #endif // TARGET_ARM || TARGET_ARM64
 
     GcInfoDecoder decoder(GCInfoToken(p), flags);
     *pRetValueKind = GetGcRefKind(decoder.GetReturnKind());
 
+#if defined(TARGET_AMD64) && defined(TARGET_OSX)
+    // Compact unwinding on macOS cannot properly handle unwinding the function prolog
+    if ((PTR_VOID)pRegisterSet->IP == pNativeMethodInfo->pMethodStartAddress)
+    {
+        *ppvRetAddrLocation = (PTR_PTR_VOID)pRegisterSet->GetSP();
+        return true;
+    }
+    else if ((PTR_UInt8)pRegisterSet->IP < (PTR_UInt8)pNativeMethodInfo->pMethodStartAddress + decoder.GetPrologSize())
+    {
+        return false;
+    }
+#endif
```

Comment:
W.R.T. arm64 prologs and unwind:

I checked the docs and the codegen. There's five different prologs that could be generated. Basically all of them adjust the `SP` pointer in a way that is incompatible with compact unwinding. There may be a theoretical edge case where no registers are saved except for LR and FP which would match the frame-based pattern. In practice I didn't see this happen on the unit tests that I examined with `xcrun unwinddump`, they always produced the DWARF fallback encoding.

We would definitely need to revisit this if the frame layout changes and we start emitting the compact unwinding.

Reply:
Not being able to unwind in prologs is a concern. We cannot unwind in epilogs either. It might make it hard to hijack loops that call small methods.

Reply:
There are different ways to approach it. We can handle known prolog types generated by the JIT. We can also tell the linker not to generate compact unwinding. There are drawbacks to each of these options. This is just minimal hack to see if the test will still crash or not. While I tried it locally on a small sample I think it's worth running on the CI machine.

Reply:
So, it looks like ARM64 is actually ok with prologs, and on AMD64 epilogs are ok (we do not need to unwind them). Not ideal, but better than I understood originally

Reply:
> We can also tell the linker not to generate compact unwinding

How bad can it get? If that is no worse than Linux, it might be acceptable.

Reply:
It should be possible to handle AMD64 prologs in similar way to the epilogs (checking known subset of instruction sequences). If this stops crashing the tests then at least we have isolated the problem to much smaller scope.

Reply:
For the preventing crashes due to this, I think the current approach is ok - we know the size of the prolog, so we can just not hijack there for now.

Reply:
> How bad can it get? If that is no worse than Linux, it might be acceptable.

Hard to say. I tried it on my machine and it seemed to avoid the crashes. It felt subjectively way more sluggish so possibly there are some performance and/or size pitfalls. Having tested it on Rosetta emulation on overheating laptop definitely didn't help though. I am fine with any stopgap solution as long as it resolves the crashes and we understand what is going on.

Reply:
If I understand it correctly, compact info is a size win (compared to DWARF), and without it thigs become the same as on Linux. That is not too bad.

Form hijacking side of view, with DWARF we can hijack in any place in a method, except ARM64 epilogs, which tend to be tiny. That is close to "as good as it gets".

So if the tradeoff is really just the exe size being like on Linux, it would be acceptable. We can look to improve this later.

Reply:
Let's see what we will get form the lab run with the current workaround. We may have other issues. 

We should consider disabling compact unwind though, if that is just a modest size increase.

Reply:
The compact unwinding is both size and speed win. It has more efficient lookup table to locate info for a given method address. Interpreting the compact info is also somewhat faster when doing exception unwinding on a longer stack trace. The drawback is that it doesn't handle prologs/epilogs. In theory some of the encodings could handle it (especially on ARM64 but also lot of cases on AMD64) but since no one really tried to interpret it that way I didn't even consider it.

On AMD64 we can choose not to generate/use the compact unwinding tables. On ARM64 they are mandatory and the linker breaks if you don't generate it. It also needs to be interpreted in order to be able to unwind over native frames (eg. for exceptions thrown from reverse P/Invoke).